packages/steelpenny/src/lib/controllers/dataexchange/request.js

```
@@ -54,4 +56,200 @@ const getStatistics = (bookshelf, query) => {
   });
 };
 
-module.exports = { loadFilterableModalities, getStatistics };
+/**
+ * Make a tree from a flat structure.
+ *
+ * @todo This passes all `items` down recursively. Improve performance by only
+ * passing down needed items, cutting down on `Array#reduce` loops.
+ *
+ * @param {Object[]} items Expected to be pre-sorted, with the parent as the
+ * first item.
+ * @param {Object} [parentItem]
+ * @returns {Object}
+ */
+const _makeTree = (items, parentItem) => {
+  const localParentItem = parentItem || items[0];
+  const _children = items.reduce((memo, item) => { // eslint-disable-line arrow-body-style
+    return item._parentId === localParentItem._id ?
+      memo.concat(_makeTree(items, item)) :
```

We need a depth limit to recursion. You should throw an application error. @drewlandis believes that 100 groups/filters is a reasonable limit.